### ISSUE TITLE
onClick firing for disabled buttons in v15.4.0

### ISSUE THREAD

**Issue by RoccoC** [facebook/react#8308]
With the latest 15.4.0 ReactJS release, `onClick` is now firing for disabled `button` elements which have child HTML elements:

https://jsfiddle.net/rcataldo/69z2wepo/62589/

In the previous ReactJS version, the onClick handler would not be called.

Note that onClick _is not called_ if I remove the `<span>` from the button.

**Comment by gaearon** [comment]
cc @spicyj @

**Comment by nhunzaker** [comment]
Is the target element the `span` or `button`? I can take a deeper look later tonight, but my first hunch is that we need to look at the way `SimpleEventPlugin` filters out disabled elements.

`SimpleEventPlugin` and bails out if the target element is disabled, but I don't know that we account for cases where the target element lives inside of a disabled element.

```
<button disabled>
    <span onClick />    <------ Click!
</button>
```

So if the target is the `span`, I wonder if it's skipping past our disabled element filter, leading to a two phased dispatch (bubbling and capturing) that bubbles up from the `span`.

The result would be an event dispatch on both the button _and_ the span.

Again, purely speculative, but maybe it can help out @spicyj until I can look at it in a few hours.

**Comment by RoccoC** [comment]
The target is the `button` element:

```
<button disabled onClick={() => { alert("Click!"); }}>
  <span>My Button</span>
</button>
```

The same but without the `span` works as expected and does not dispatch the event on the `button`:

```
<button disabled onClick={() => { alert("Click!"); }}>
  My Button
</button>
```

**Comment by nhunzaker** [comment]
@RoccoC :+1: React manipulates the `target` field as it bubbles events up the component tree. I wonder if we can performantly crawl up the tree to see if an element is disabled.

**Comment by RoccoC** [comment]
I see the function `shouldPreventMouseEvent` in `SimpleEventPlugin` which seems to handle whether a mouse event should be prevented based on component instance's tag and `disabled` prop, but the SimpleEventPlugin's `extractEvents` function is passing the event target (in this case the `span`) to `shouldPreventMouseEvent`, not the actual dispatch target to which the listener is assigned (e.g. the `button`).

Adding this to the end of `extractEvents` fixes this, but there is probably a cleaner way:

 ```javascript
/* SimpleEventPlugin.js */
// extractEvents snip
if (shouldPreventMouseEvent(event._dispatchInstances)) {
  return null;
}
return event;
```

**Comment by Aweary** [comment]
@RoccoC `event` isn't defined until after that switch statement, when the constructor is determined and an instance is pulled from the pool. [`accumulateDirectionalDispatches`](https://github.com/facebook/react/blob/e3131c1d55d6695c2f0966379535f88b813f912b/src/renderers/shared/shared/event/EventPropagators.js#L41-L54) then eventually populates `event._dispatchInstances`.

[`traverseTwoPhase`](https://github.com/facebook/react/blob/aca2e0403691a2842d4a9d8964c9b1016e248b6a/src/renderers/shared/shared/ReactTreeTraversal.js#L93-L106) ends up calling `accumulateDirectionalDispatches` for every element in the target element's parent tree, eventually finding the disabled button's listener thanks to [`listenerAtPhase`](https://github.com/facebook/react/blob/e3131c1d55d6695c2f0966379535f88b813f912b/src/renderers/shared/shared/event/EventPropagators.js#L48). That listener is then [later executed](https://github.com/facebook/react/blob/3d48139dd7a5c3dbd9ba1814bd31d9d8d8d71eff/src/renderers/shared/shared/event/EventPluginUtils.js#L128).

At no point does any step check to see if the instance for the listener is disabled. Maybe we can do that early on in `accumulateDirectionalDispatches`?

```js
var listener = listenerAtPhase(inst, event, phase);
  if (listener && shouldExecuteListener(inst, event, phase)) {
    // only redefine `_dispatchInstances` and _dispatchListener` if it should actually execute
  }
```

**Comment by Aweary** [comment]
> Adding this to the end of extractEvents fixes this, but there is probably a cleaner way:

@RoccoC one issue there is `_dispatchInstances` and `_dispatchListeners` can be arrays, and there may be cases where some listeners should be dispatched and some shouldn't. So we'd have to introduce more logic to `shouldPreventMouseEvent` and make it polymorphic as a result, which isn't ideal.

I think it would be better if we could prevent the listener from being registered with the event in the first place.

**Comment by nhunzaker** [comment]
> I think it would be better if we could prevent the listener from being registered with the event in the first place.

@Aweary aye. Happy to dig into this unless someone has already taken it. Here's a test case:

https://github.com/nhunzaker/react/compare/master...nhunzaker:disabled-button-children

**Comment by nhunzaker** [comment]
Following @Aweary's thoughts, I first looked into avoiding the subscription all together. However, it turns out, there are some additional things to figure out there. 

When React updates a component, it adds/removes event listeners based on whether or not they have changed. This is problematic for disabling events. Changing the `disabled` prop is a separate operation from attaching/removing listeners. If we prevent disabled elements from attaching mouse events, we need to make sure they attached if an element stops being disabled.

The simplest solution does appear to be just making `SimpleEventPlugin` smarter by modifying `shouldPreventMouseEvent` to be:

```javascript
function shouldPreventMouseEvent(inst) {
  if (inst) {
    var focus = inst;

    while (focus) {
      if (focus._currentElement && focus._currentElement.props.disabled) {
        return isInteractive(focus._tag);
      }

      focus = focus._hostParent;
    }
  }

  return false;
}
```

Basically bubble up the tree and see if the child is within a disabled parent. If so, return true/false if that element is interactive or not.

---

**Side note:** it looks like there is a substantial difference in the event implementation between master and `15.x`. We might need to implement this different in both versions, but I think the solution could be the same in principle.

**Comment by nhunzaker** [comment]
I sent out this PR for discussion:
https://github.com/facebook/react/pull/8329

**Comment by joh1234** [comment]
Update README
https://goo.gl/JBljdR

**Comment by zargold** [comment]
Not sure if this was closed or not but I'm still experiencing the issue: (will check different versions):
[Codepen Example](http://codepen.io/zargold/pen/mWbKGQ)

**Comment by monkindey** [comment]
Hi, @zargold, I saw your sample, just the a tag trigger but the a tag have not the disabled property. [Element a](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a) for more reading.

**Comment by zargold** [comment]
Yes but in my ES6 code I do not have an `<a></a>` I have: 
``` babel
        <button
          className='button' disabled={true}
          onClick={(e) => addNewStory(e, 'greenStories')}
        ><span className='fa fa-plus-circle'></span>Add Story to Top
        </button>
```
but when rendered it turns into:
``` html
<a disabled>
<span className='fa fa-plus-circle'></span>Add Story to Top
</a>
```
Which did not work

**Comment by monkindey** [comment]
Sorry, I can not follow you.  I saw your ES6 code in the Codepen

```javascript
<a disabled={this.state.shouldDisable} onClick={(e) => funcCheck(e)}>
    <span>Stuff in button</span>
</a>
```
it's obviously have the A tag in the React render function.

Did you put the wrong code in the Codepen ?

**Comment by zargold** [comment]
Ok thanks monkindey but yes this was actually an issue my babel... and by restarting uninstalling babel reinstalling babel after switching from `<a>` to `<button>` it stayed a button when rendered and then worked.

**Comment by gaearon** [comment]
This has since been fixed.

**Comment by ClemDelp** [comment]
@gaearon It looks like the bug is back since react@16.0.0

**Comment by sophiebits** [comment]
If you have a case where this doesn't work properly, please open a new issue.

**Comment by bryantobing12** [comment]
not sure if this is related. 

In react when we set button props `disabled`, the onClick event will not fire.
When the user enables the button from devtool/inspect-element by removing the disabled attribute. the onClick still will not fire.
I have tried this and that's work and this is my expected behavior. because I want to prevent users from clicking by role access to the button.

I'm just still not certain that this is correct.
